lib/attrsets.nix

```diff
@@ -135,14 +135,35 @@ rec {
        filterAttrsRecursive (n: v: v != null) { foo = { bar = null; }; }
        => { foo = {}; }
   */
-  filterAttrsRecursive = pred: set:
+  filterAttrsRecursive = filterAttrsRecursiveCond (x: true);
+
+  /* Like `filterAttrsRecursive', but it takes an additional predicate
+     function that tells it whether to recurse into an attribute
+     set.  If it returns false, `filterAttrsRecursiveCond' does not
+     recurse, but does apply the filter function.  If it returns true, it
+     does recurse, and does not apply the filter function.
+
+     Type:
+       filterAttrsRecursiveCond ::
+         (AttrSet -> Bool) -> (String -> a -> Bool) -> AttrSet -> AttrSet
+
+     Example:
+       # To prevent recursing into derivations (which are attribute
+       # sets with the attribute "type" equal to "derivation"):
+       filterAttrsRecursiveCond
+         (as: !(as ? "type" && as.type == "derivation"))
+         (x: ... do something ...)
+         attrs
+  */
+  filterAttrsRecursiveCond = cond: pred: set:
     listToAttrs (
       concatMap (name:
         let v = set.${name}; in
         if pred name v then [
```

Comment:
This could be lib.optional.

Reply:
Some functions in `lib` affect performance a lot, so it makes sense to lean towards plain expressions instead of simple utility functions, to avoid constructing more thunks, closures, envs, etc, than necessary.